File: Project/MyProject/MyProject/Source/MyProject/Private/BasicFrame/Libs/DelayHandle/DelayNoOrPriorityHandleMgr.cpp
#include "MyProject.h"
#include "DelayNoOrPriorityHandleMgr.h"
#include "INoOrPriorityObject.h"
#include "ITickedObject.h"
#include "IDelayHandleItem.h"
#include "MClassFactory.h"
#include "MacroDef.h"
#include "Ctx.h"
#include "LogSys.h"
#include "LogTypeId.h"

MY_BEGIN_NAMESPACE(MyNS)

M_IMPLEMENT_AND_REGISTER_CLASS(DelayNoOrPriorityHandleMgr, DelayNoOrPriorityHandleMgrBase)

DelayNoOrPriorityHandleMgr::DelayNoOrPriorityHandleMgr()
{

}

void DelayNoOrPriorityHandleMgr::init() 
{
	Super::init();
}

void DelayNoOrPriorityHandleMgr::dispose()
{
	Super::dispose();
}

bool DelayNoOrPriorityHandleMgr::isEmpty()
{
	bool ret = (this->mNoOrPriorityList.count() == 0);
	return ret;
}

void DelayNoOrPriorityHandleMgr::addNoOrPriorityObject(INoOrPriorityObject* priorityObject, float priority)
{
	this->_addObject((IDelayHandleItem*)priorityObject, priority);
}

void DelayNoOrPriorityHandleMgr::removeNoOrPriorityObject(ITickedObject* tickObj)
{
	this->_removeObject((IDelayHandleItem*)tickObj);
}

void DelayNoOrPriorityHandleMgr::_addObject(IDelayHandleItem* delayObject, float priority)
{
	if (nullptr != delayObject)
	{
		if (this->_isInDepth())
		{
			Super::_addObject(delayObject, priority);
		}
		else
		{
			if (!this->mNoOrPriorityList.contains((INoOrPriorityObject*)delayObject))
			{
				this->mNoOrPriorityList.addNoOrPriorityObject((INoOrPriorityObject*)delayObject, priority);
			}
		}
	}
	else
	{
		if (MacroDef.ENABLE_LOG)
		{
			GLogSys->log("DelayPriorityHandleMgr::_addObject, failed", LogTypeId::eLogCommon);
		}
	}
}

void DelayNoOrPriorityHandleMgr::_removeObject(IDelayHandleItem* delayObject)
{
	if (nullptr != delayObject)
	{
		if (this->_isInDepth())
		{
			Super::_removeObject(delayObject);
		}
		else
		{
			if(this->mNoOrPriorityList.contains((INoOrPriorityObject*)delayObject))
			{
				this->mNoOrPriorityList.removeNoOrPriorityObject((INoOrPriorityObject*)delayObject);
			}
		}
	}
	else
	{
		if (MacroDef.ENABLE_LOG)
		{
			GLogSys->log("DelayPriorityHandleMgr::_removeObject, failed", LogTypeId::eLogCommon);
		}
	}
}

MY_END_NAMESPACE

File: Project/MyProject/MyProject/Content/MyAsset/Lua/MyLua/Libs/DelayHandle/DelayNoOrPriorityHandleMgr.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.Core.GObject");

MLoader("MyLua.Libs.DelayHandle.DelayNoOrPriorityHandleMgrBase");

-- 延迟优先级处理管理器
local M = GlobalNS.Class(GlobalNS.DelayNoOrPriorityHandleMgrBase);
M.clsName = "DelayNoOrPriorityHandleMgr";
GlobalNS[M.clsName] = M;

function M:ctor()

end

function M:init()
	M.super.init(self);
end

function M:dispose()
	self.mNoOrPriorityList:clear();
	
	M.super.dispose(self);
end

function M:addObject(delayObject, priority)
	if (nil ~= delayObject) then
		if (self:isInDepth()) then
			M.super.addObject(self, delayObject, priority);
		else
			self.mNoOrPriorityList:addNoOrPriorityObject(delayObject, priority);
		end
	else
		if (MacroDef.ENABLE_LOG) then
			GCtx.mLogSys:log("DelayPriorityHandleMgr::addObject, failed", GlobalNS.LogTypeId.eLogCommon);
		end
	end
end

function M:removeObject(delayObject)
	if (nil ~= delayObject) then
		if (self:isInDepth()) then
			M.super.removeObject(self, delayObject);
		else
			self.mNoOrPriorityList:removeNoOrPriorityObject(delayObject);
		end
	else
		if (MacroDef.ENABLE_LOG) then
			GCtx.mLogSys:log("DelayPriorityHandleMgr::removeObject, failed", GlobalNS.LogTypeId.eLogCommon);
		end
	end
end

function M:addNoOrPriorityObject(priorityObject, priority)
	self:addObject(priorityObject, priority);
end

function M:removeNoOrPriorityObject(tickObj)
	self:removeObject(tickObj);
end

return M;